text/3874-build-std-always.md

```diff
@@ -0,0 +1,1543 @@
+
+By minimising the differences between a newly-built std and a pre-built std,
+there is less chance of the user experiencing bugs or unexpected behaviour from
+the well-tested and supported pre-built std. Later proposals will extend the
+`build-std` option with customised standard library builds that use the user's
+profile.
+
+↩ [*Proposal*][proposal]
+
+## Why add `build-std.crates`?
+[rationale-build-std-crate]: #why-add-build-stdcrates
+
+Not all standard library crates will build on all targets. In a `no_std` project
+for a tier three target, `build-std.crates` gives the user the ability to limit
+which crates are built to those they know they need and will build successfully.
+
+*See [Standard library dependencies*][rfcs#3875] for an alternative to
+`build-std.crates`.*
+
+↩ [*Proposal*][proposal]
+
+## Why use the lockfile of the `rust-src` component?
+[rationale-lockfile]: #why-use-the-lockfile-of-the-rust-src-component
+
+Using different dependency versions for the standard library would invalidate
+the upstream testing of the standard library. In particular, some crates use
+unstable APIs when included as a dependency of the standard library meaning that
+there is a high risk of build breakage if any package version is changed.
+
+Using the lockfile included in the `rust-src` component guarantees that the same
+dependency versions are used as in the pre-built standard library. As the
+standard library does not re-export types from its dependencies, this will not
+affect interoperability with the same dependencies of different versions used by
+the user's crate.
+
+Using the lockfile does prevent Cargo from resolving the standard library
+dependencies to newer patch versions that may contain security fixes. However,
+this is already impossible with the pre-built standard library.
+
+See
+[*Why vendor the standard library's dependencies?*][rationale-vendoring]
+
+↩ [*Proposal*][proposal]
+
+## Why not build the standard library in incremental?
+[rationale-incremental]: #why-not-build-the-standard-library-in-incremental
+
+The standard library sources are not intended to be modified locally, similarly
+to those Cargo fetches from `registry` or `git` sources. Incremental compilation
+would only add a compilation time overhead for any package sources which do not
+change.
+
+↩ [*Proposal*][proposal]
+
+## Why not produce a `dylib` for the standard library?
+[rationale-no-dylib]: #why-not-produce-a-dylib-for-the-standard-library
+
+The standard library supports being built as both a `rlib` and a `dylib` and
+both are shipped as part of the `rust-std` component. As the `dylib` does not
+contain a metadata hash, it can be rebuilt unnecessarily when toolchain versions
+change (e.g. switching between stable and nightly and back). The `dylib` is only
+linked against when `-Cprefer-dynamic` is used. build-std will initially be
+conservative and not include the `dylib` and `-Cprefer-dynamic` would fail
+compilation.
```

Comment:
I'm pretty sure `-Cprefer-dynamic` would fallback to static linking rather than failing compilation.